src/microwasm.rs

```diff
@@ -0,0 +1,1696 @@
+        }
+    }
+}
+
+impl Value {
+    fn default_for_type(ty: SignlessType) -> Self {
+        match ty {
+            Type::Int(Size::_32) => Value::I32(0),
+            Type::Int(Size::_64) => Value::I64(0),
+            Type::Float(Size::_32) => Value::F32(Ieee32(0)),
+            Type::Float(Size::_64) => Value::F64(Ieee64(0)),
+        }
+    }
+}
+
+/// Whether to interpret an integer as signed or unsigned
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Signedness {
+    Signed,
+    Unsigned,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Size {
+    _32,
+    _64,
+}
+
+type Int = Size;
+type Float = Size;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SignfulInt(Signedness, Size);
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Type<I> {
+    Int(I),
+    Float(Size),
+}
+
+impl fmt::Display for SignfulType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Type::Int(i) => write!(f, "{}", i),
+            Type::Float(Size::_32) => write!(f, "f32"),
+            Type::Float(Size::_64) => write!(f, "f64"),
+        }
+    }
+}
+
+impl fmt::Display for SignlessType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Type::Int(Size::_32) => write!(f, "i32"),
+            Type::Int(Size::_64) => write!(f, "i64"),
+            Type::Float(Size::_32) => write!(f, "f32"),
+            Type::Float(Size::_64) => write!(f, "f64"),
+        }
+    }
+}
+
+impl fmt::Display for SignfulInt {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SignfulInt(Signedness::Signed, Size::_32) => write!(f, "i32"),
+            SignfulInt(Signedness::Unsigned, Size::_32) => write!(f, "u32"),
+            SignfulInt(Signedness::Signed, Size::_64) => write!(f, "i64"),
+            SignfulInt(Signedness::Unsigned, Size::_64) => write!(f, "u64"),
+        }
+    }
+}
+
+pub type SignlessType = Type<Size>;
+pub type SignfulType = Type<SignfulInt>;
+
+pub const I32: SignlessType = Type::Int(Size::_32);
+pub const I64: SignlessType = Type::Int(Size::_64);
+pub const F32: SignlessType = Type::Float(Size::_32);
+pub const F64: SignlessType = Type::Float(Size::_64);
+
+pub mod sint {
+    use super::{Signedness, SignfulInt, Size};
+
+    pub const I32: SignfulInt = SignfulInt(Signedness::Signed, Size::_32);
+    pub const I64: SignfulInt = SignfulInt(Signedness::Signed, Size::_64);
+    pub const U32: SignfulInt = SignfulInt(Signedness::Unsigned, Size::_32);
+    pub const U64: SignfulInt = SignfulInt(Signedness::Unsigned, Size::_64);
+}
+
+pub const SI32: SignfulType = Type::Int(sint::I32);
+pub const SI64: SignfulType = Type::Int(sint::I64);
+pub const SU32: SignfulType = Type::Int(sint::U32);
+pub const SU64: SignfulType = Type::Int(sint::U64);
+pub const SF32: SignfulType = Type::Float(Size::_32);
+pub const SF64: SignfulType = Type::Float(Size::_64);
```

Comment:
I could have `mod signful`/`mod signless` with constants that have overlapping names, but unfortunately I can't have a generic constant/generic constant function that can return either since Rust doesn't support const fns in traits. The prefix is because we have "signless" and "signful" versions.